Xfem cutter using crack tip stress field

lynnmunday opened the pull request:
closes #27659

lynnmunday committed:
adding vector postprocessor to compute the volume averaged stress at a crack tip

lynnmunday committed:
add new method to return crack plane normals to otehr objects

lynnmunday committed:
use crack tip stress in MeshCut2dFractureUserObject to grow crack.

lynnmunday committed:
add tests for crackFrontNormalStress ref #27659

lynnmunday committed:
Refactored CrackFrontDefinition to always use a vector to store crack plane normals.

Also fixed xfem tests file with wrong input file.

lynnmunday committed:
Add missing information to tests file.

lynnmunday committed:
adding documentation for CrackFrontNormalStress

lynnmunday committed:
added documentation and testing for crack front stress to solid_mechanics module.

lynnmunday committed:
General clean up  -- All solid_mechanics and xfem tests pass
Being more consistant with getting crack plane normals in CrackFrontDefinition.
Removed Id or position from crack stress.  It was not providing anything useful. This required several tests to be regolded.

lynnmunday committed:
Updating MeshCut2DFractureUserObject documentation for CrackFrontNormalStress

lynnmunday committed:
Defaulting execution flag for crackFrontStress to timestep begin and nonlinear.
Without timestep begin, it will not run for tests with no nonlinear solves.
Without nonlinear, it will not update during xfem_updates
Adding test for stress driven cutting and k driven cutting and a combination of both
moving xfem 2d mesh cutter tests to their own directory
closes #27659

lynnmunday committed:
Fixing error in orientation of crackFrontStress box and regolding files, Fixing file_base naming in xfemMeshCutter2D output.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2251713/) on 56d371e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27709/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2251757/) on 56d371e wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>be45bd</th>
      <th colspan=4>#27709 56d371</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.86%</td>
      <td>84.89%</td>
      <td>+0.03%</td>
      <td>91.18%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>27679</td>
      <td>27752</td>
      <td>+73</td>
      <td>93</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4938</td>
      <td>4938</td>
      <td>-</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27709/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27709/coverage/solid_mechanics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>be45bd</th>
      <th colspan=4>#27709 56d371</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.37%</td>
      <td>82.38%</td>
      <td>+0.01%</td>
      <td>85.42%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6882</td>
      <td>6904</td>
      <td>+22</td>
      <td>41</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1473</td>
      <td>1477</td>
      <td>+4</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27709/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27709/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27709/coverage/xfem)

</details>

## Warnings

- `xfem` new line coverage rate 85.42% is less than the suggested 90.0%



This comment will be updated on new commits.

bwspenc commented on modules/solid_mechanics/src/vectorpostprocessors/CrackFrontNormalStress.C:
```diff
@@ -0,0 +1,159 @@
+                                          "The CrackFrontDefinition user object name");
+  params.addRequiredParam<Real>("box_length", "Distance in front of crack front.");
+  params.addRequiredParam<Real>("box_height", "Distance normal to front of crack front.");
+  params.addParam<Real>("box_width", 1, "Distance tangent to front of crack front.");
```
For 3D, it seems like it might be better to use the spacing of points along the crack front for the width. Let's discuss.

lynnmunday replied:
I did not fix this in this pr.

bwspenc commented on modules/xfem/doc/content/source/userobjects/MeshCut2DFractureUserObject.md:
```diff
@@ -4,19 +4,31 @@
+
+where $K_I$ and $K_{II}$ are the mode I and II stress intensity factors provided by the fracture integral and the material property $K_c$ is given defined in the input file by [!param](/UserObjects/MeshCut2DFractureUserObject/k_critical).  The crack growth direction is given by the direction the maximized the crack-tip hoop stress, given by Equation 5 in [!cite](jiang2020).  The growth increment  ...
+
+Near a free surface, the integration volume of the q-function used to compute the `InteractionIntegral` will intersect the surface, leading to a reduction in the fracture integral values.  This can lead to cracks becoming stuck near free surfaces.  For these cases, a maximum stress criterion computed using [CrackFrontNormalStress.md] vectorpostprocessor can be used for crack growth using an addit ...
```
```suggestion
Near a free surface, the integration volumes of the rings used to compute the `InteractionIntegral` will intersect the surface, leading to a reduction in the fracture integral values.  This can lead to cracks becoming unable to farther propagate as they approach free surfaces.  For these cases, a maximum stress criterion computed using [CrackFrontNormalStress.md] vectorpostprocessor can be used for crack growth using an additional failure criterion given by:
```

bwspenc commented on modules/xfem/test/tests/mesh_cut_2D_fracture/tests:
```diff
@@ -0,0 +1,43 @@
+  design = 'MeshCut2DFunctionUserObject.md'
+  [2D]
+    requirement = 'The XFEM module shall represent a propagating crack in a 2D mechanics model with XFEM, '
+    'where the crack is defined using a topologically surface cutting mesh by MeshCut2DFunctionUserObject'
```
```suggestion
    'where the crack is defined using a topologically defined surface cutting mesh by MeshCut2DFunctionUserObject'
```

bwspenc commented on modules/xfem/test/tests/mesh_cut_2D_fracture/tests:
```diff
@@ -0,0 +1,43 @@
+                'stress_check_CrackFrontNormalStress_0004.csv'
+      # XFEM requires --enable-unique-ids in libmesh
+      unique_id = true
+      detail = ' and crack growth is defined by a function with the average crack tip stress changes with crack orientation '
+               'and properly accounts averaging volume intersecting boundary.'
```
Proofread this -- it's not quite clear to me what you mean.

bwspenc commented on modules/xfem/include/userobjects/MeshCut2DFractureUserObject.h:
```diff
@@ -50,4 +39,27 @@ class MeshCut2DFractureUserObject : public MeshCut2DUserObjectBase
+  const Real _stress_threshold;
+
+  /// dummy vectors for vpps to point to if not defined(these are not const because reporters can change their size)
+  std::vector<Real> _zeros_vec;
```
Why didn't we need this before?

lynnmunday replied:
changed to pointers

bwspenc commented on modules/xfem/include/userobjects/MeshCut2DFractureUserObject.h:
```diff
@@ -50,4 +39,27 @@ class MeshCut2DFractureUserObject : public MeshCut2DUserObjectBase
+  /// dummy vectors for vpps to point to if not defined(these are not const because reporters can change their size)
+  std::vector<Real> _zeros_vec;
+
+  /// References fracture integral ki if input
```
"if computed" instead of "if input"?

bwspenc commented on modules/xfem/include/userobjects/MeshCut2DUserObjectBase.h:
```diff
@@ -62,7 +66,7 @@ class MeshCut2DUserObjectBase : public GeometricCutUserObject
   /**
    * This vector of pairs orders crack tips to make the order used in this class the same as those
-   * for the fracture integrals vectorpostProcessorsin created by CrackFrontDefinition.
+   * for the fracture integrals vectorpostProcessors created by CrackFrontDefinition.
```
```suggestion
   * for the fracture integrals VectorPostprocessors created by CrackFrontDefinition.
```

bwspenc commented on modules/xfem/src/userobjects/MeshCut2DFractureUserObject.C:
```diff
@@ -63,40 +82,60 @@ MeshCut2DFractureUserObject::initialize()
+               _original_and_current_front_node_ids.size());
+
+  if (_use_stress && ((_stress_vpp.size() != _original_and_current_front_node_ids.size())))
+    mooseError("stress_vectorpostprocessor should have the same number of crack tips as "
```
```suggestion
    mooseError("stress_vectorpostprocessor should have the same number of crack front points as "
```

bwspenc commented on modules/solid_mechanics/doc/content/source/vectorpostprocessors/CrackFrontNormalStress.md:
```diff
@@ -0,0 +1,24 @@
+# CrackFrontNormalStress
```
Let's rename to `CrackFrontNonlocalStress`

bwspenc commented on modules/solid_mechanics/include/vectorpostprocessors/CrackFrontNormalStress.h:
```diff
@@ -0,0 +1,72 @@
+   * @param current_node Node at which q is evaluated
+   * @return q
+   */
+  Real CrackFrontBox(std::size_t crack_front_point_index, const Point & qp_coord) const;
```
IsPointInBox, BoxWeightingFunction

bwspenc commented on modules/solid_mechanics/src/userobjects/CrackFrontDefinition.C:
```diff
@@ -1665,7 +1666,7 @@ CrackFrontDefinition::createQFunctionRings()
     }
   }
-  else // 3D: The q-function defines one integral domain around each crack front node
+  else // _use_mesh_cutter: The q-function defines one integral domain around each crack front node
```
Just remove the `_use_mesh_cutter` bit.

bwspenc commented on modules/xfem/src/userobjects/MeshCut2DFractureUserObject.C:
```diff
@@ -63,40 +82,60 @@ MeshCut2DFractureUserObject::initialize()
-  // InteractionIntegral vpp
-  if (k1.empty())
+  // InteractionIntegral or crackFrontStress vpp
+  if (_ki_vpp.empty() && _stress_vpp.empty())
```
If you're hitting this, it seems like these aren't getting run in the right order. Let's check that.

lynnmunday committed:
Apply suggestions from code review

Co-authored-by: Ben Spencer <benjamin.spencer@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2254300/) on ebd17cc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27709/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27709/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4409da00956c2511d3f6c6db886956a265cd4dd9`

lynnmunday committed:
fixing reviewer comments,  Renaming object to CrackFrontNonlocalStress and small typos

lynnmunday committed:
adding crack front position to crackFrontStress output and regolding files.

lynnmunday committed:
Use a pointer to get VPP values in crackFrontNonlocalStress.

Get xfem_flag in domainIntegralAction as a way to test whether solid_mechanics is being built with xfem.
This will then trigger different flags to be set by the domainIntegralAction for solid_mecahnics vpps used by xfem

Added some comments on in DomainIntegralAction about flags that should be set synchronize the execution of xfem meshcutter userobject, solid mechanics vpps and solid_mechanics crackPointDefinition userobject.

lynnmunday committed:
fixing error in how cracks were offset when they nucleate on the boundary.   The previous version did not move the crack close to teh boundary and therefor was not nucleating the full length of the crack provided by the user

lynnmunday commented on modules/xfem/src/userobjects/MeshCut2DRankTwoTensorNucleation.C:
```diff
@@ -175,15 +175,22 @@ MeshCut2DRankTwoTensorNucleation::doesElementCrack(
     // equal to nucleation_length
     if (is_point_0_on_external_boundary)
     {
+      // first move point back to edge
```
@bwspenc this is an error I found.  This is now correct and I ran some a new test case that is in our milestone to check it.  See the second figure in this section:  XFEM Crack Nucleation in the Bulk
Without this commit, the edge cracks were sticking way out of the domain.

lynnmunday commented:
@bwspenc I have all of the tests passing and addressed all of your comments except for the two that are not marked resolved.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2257070/) on 206db70 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2257070/) on 206db70 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2257070/) on 206db70 : invalidated by @bwspenc

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2257070/) on 206db70 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)

lynnmunday commented:
@bwspenc Will you review and merge?  Its now passing all tests.

bwspenc commented on modules/solid_mechanics/src/actions/DomainIntegralAction.C:
```diff
@@ -373,11 +388,11 @@ DomainIntegralAction::act()
-      params.set<ExecFlagEnum>("execute_on") = {EXEC_INITIAL, EXEC_TIMESTEP_END, EXEC_NONLINEAR};
       // The CrackFrontDefinition updates the vpps and MUST execute before them
       params.set<int>("execution_order_group") = -1;
+      params.set<ExecFlagEnum>("execute_on") = xfem_exec_flags;
```
I think we need `EXEC_INITIAL` here, and that's not in `xfem_exec_flags`. I'd like to talk about the lingering issues noted in the TODO above too.

lynnmunday replied:
This is trying handle the order of execution for the vpps and user_objects.  We wanted to handle this in a separate PR.  @dschwen was going to help us get these straightened out in a separate PR.

bwspenc replied:
Yeah, I get that. I just see that `EXEC_INITIAL` was in the list of exec flags, but it's not in the list defined on line 378. I think maybe the fix is to just add `EXEC_INITIAL` to `xfem_exec_flags` on line 378. Apparently we don't have tests that catch it, but I do recall that this was important.

bwspenc commented on modules/solid_mechanics/src/vectorpostprocessors/CrackFrontNonlocalStress.C:
```diff
@@ -0,0 +1,165 @@
+                                          "The CrackFrontDefinition user object name");
+  params.addRequiredParam<Real>("box_length", "Distance in front of crack front.");
+  params.addRequiredParam<Real>("box_height", "Distance normal to front of crack front.");
+  params.addParam<Real>("box_width", 1, "Distance tangent to front of crack front.");
```
Weren't we going to have this default to something based on the segment length? Regardless, a default of 1 probably isn't a good idea here.

bwspenc replied:
Thanks, I missed where the error was generated. I think the best way to handle this is to just not set the default of 1 in line 32. It doesn't really make much (maybe any) difference in the way the code runs, but it wouldn't print 1 as a default value in the documentation. There's not a case where you would fall back on using 1, so it's not meaningful to set a default value.

bwspenc commented on modules/xfem/src/userobjects/MeshCut2DFractureUserObject.C:
```diff
@@ -26,25 +26,43 @@ MeshCut2DFractureUserObject::validParams()
-    _ring_number_string(std::to_string(getParam<unsigned int>("ring_number")))
+    _use_k(isParamValid("k_critical")),
+    _use_stress(isParamValid("stress_threshold")),
+    _k_critical(_use_k ? getParam<Real>("k_critical") : 1.0e12),
```
I assume the 1e12 in this line and the one below are "big numbers". We should use `std::numeric_limits<Real>::max()` instead.

bwspenc commented on modules/xfem/src/userobjects/MeshCut2DUserObjectBase.C:
```diff
@@ -312,6 +324,10 @@ MeshCut2DUserObjectBase::growFront()
       Point new_node_offset = direction_iter->second;
       Point x = this_point + new_node_offset;
 
+      // FIXME LYNN check if line between this_point and x crosses an element in the cutter mesh or
```
Is this still relevant?

lynnmunday committed:
addressing review comments.  closes #27659

moosebuild commented:
Job [Public app tests](https://civet.inl.gov/job/2326730/) on 8454969 : invalidated by @bwspenc

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/2326723/) on 8454969 : invalidated by @bwspenc

moosebuild commented:
Job [Documentation public apps](https://civet.inl.gov/job/2326698/) on 8454969 : invalidated by @bwspenc
